### ISSUE TITLE
508-defect-1 [SCREENREADER]: Feedback button is difficult to access for keyboard users and effectively invisible for screen reader users

### ISSUE THREAD

**Issue by joshkimux** [department-of-veterans-affairs/va.gov-team#26017]
# [508-defect-1 :exclamation: Launchblocker](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-1)

## Feedback framework

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Definition of done

1. Review and acknowledge feedback.
1. Fix and/or document decisions made.
1. Accessibility specialist will close ticket after reviewing documented decisions / validating fix.

## Point of Contact

**VFS Point of Contact:** Angela and Josh

## User Story or Problem Statement

As a screen reader user, I expect content to be contained in landmarks so I can find them. 
As a screen reader user, I expect the feedback button to be contained within the `main` because I want to leave feedback for the main content of the page.
As a keyboard user, I expect to be able to tab to the feedback button after I've finished navigating the content on the page.

### Keyboard users 
The feedback is the final tab stop in the DOM making it highly inaccessible to keyboard users who will need to tab through the entire page and footer to reach. Its floating/sticky characteristic may be misleading to keyboard users who may think they can access it sooner since it's constantly visible.

https://user-images.githubusercontent.com/14154792/121936795-ff885a00-cd17-11eb-8843-629cd5296850.mov

Contrast this with gov.uk's feedback buttons which are located in a consistent location and can be accessed once page content ends before the footer.

https://user-images.githubusercontent.com/14154792/121937164-64dc4b00-cd18-11eb-9a0f-75887ff1ee8f.mov

### Screen reader users
If the feedback button is providing feedback to the main content of the page, it should be housed within the `main` landmark. Otherwise, it runs the risk of being undiscovered due to folks not exploring _past_ the main content of the page itself to find an option for providing feedback. 

More critically, it also isn't located within any landmark. When content is not contained in a landmark, it will be unreachable using landmark navigation, which some screen reader users prefer to use as their main method of exploring a page.

## Steps to Recreate

1. Enter `https://www.va.gov/health-care/refill-track-prescriptions/` in browser
2. Tab until you reach the feedback button, confirm it's beyond the footer
3. Inspect the feedback button, confirm that it's not contained within a landmark

## Proposed Solution

Don't make the feedback button sticky. Place it in a consistent location at the end of the `main` content similar to gov.uk.

## WCAG or Vendor Guidance (optional)

[1.3.1 Info and Relationships](https://www.w3.org/TR/WCAG20/#content-structure-separation-programmatic)

**Comment by jwoodman5** [comment]
@1Copenut Just confirming this is a defect-2, not a defect-1? We are allocating resources for these and the question came up for clarification. Thanks!

**Comment by 1Copenut** [comment]
@jwoodman5 Yes, let's treat this as a defect two. It's a lot of keypresses for users to get to the Feedback button via keyboard, so very difficult but workable. I'd consider this a high two, maybe the first two fixed after the 1's are cleared.

**Comment by jwoodman5** [comment]
Thank you for that confirmation and description, @1Copenut.

**Comment by holdenhinkle** [comment]
PR for removing the Feedback button and appending it to the `content` section, and displaying it inline (so it doesn't stick) - https://github.com/department-of-veterans-affairs/vets-website/pull/17870/files

The design needs to be updated. It doesn't look so great.

Here's a movie (skip the first few seconds). It shows tabbing before and after, and opening and closing the form before and after:

https://user-images.githubusercontent.com/9326247/125699879-0f2c64aa-0ff2-4d05-a454-e05ee0e1c7e3.mov

**Comment by holdenhinkle** [comment]
Trying to figure out where this `js` should live - https://dsva.slack.com/archives/C01CHAY3ULR/p1626353282310600

@1Copenut , when you have a chance, can you go to [this page](https://staging.va.gov/change-address/) and pop this `js` in your console to confirm this does what you want it to do?

Some comments:
I need some design direction. This doesn't look right. And the button needs to have better placement.

```js
  const main = document.querySelector('main');
  const feedbackButtonContainer = document.getElementById('kampyleButtonContainer');

  const updateFeedbackButtonInlineCss = () => {
    const feedbackButtonDiv = document.getElementById('nebula_div_btn');
    feedbackButtonDiv.style.zIndex = null;
    feedbackButtonDiv.style.position = null;
  }

  const detectButtonAdded = function(mutationsList) {
    for (let mutation of mutationsList) {
      if (mutation.addedNodes.length > 0) {
        mutation.addedNodes.forEach(node => {
          if (node.id === 'nebula_div_btn') {
            updateFeedbackButtonInlineCss()
          }
        });
      }
    }
  };

  feedbackButtonContainer.remove();
  main.append(feedbackButtonContainer);
  updateFeedbackButtonInlineCss();
  
  const observer = new MutationObserver(detectButtonAdded)
  const observeConfig = { attributes: false, childList: true };
  observer.observe(feedbackButtonContainer, observeConfig);
```

**Comment by joshkimux** [comment]
@holdenhinkle and @jwoodman5 

In terms of button placement, would it possible to unfloat the button and make it static to the bottom of the `main`?
It'll be easier for Veterans with cognitive and reading disabilities to read if it isn't rotated. More importantly, it will resolve the defect of it interfering with mobile screen readers which can render entire pages unusable.

**Comment by holdenhinkle** [comment]
@joshkimux Yes, this script un-floats the button and places it at the bottom of `main`. We definitely need a new graphic--a horizontal button--and maybe some additional design direction.

**Comment by jwoodman5** [comment]
@holdenhinkle Talked with @ianMcCullough-ob about this and we are good with the concept. Ian will drop in a new button and provide direction.

**Comment by jwoodman5** [comment]
@holdenhinkle If you read my previous comment, which I deleted, please disregard. We are good with updating the screenreader order, but actually moving the button is a larger/ongoing conversation with leadership and we need to hold off on that.

**Comment by holdenhinkle** [comment]
Okay, thanks. I can modify this PR so it only moves the button into `main` but doesn't visually move the button, if that's what you mean. Then we can address moving the button visually (un-floating it, etc.) later in a separate PR if leadership wants that.

**Comment by joshkimux** [comment]
@jwoodman5 Is there any way we can help with that ongoing conversation?

**Comment by jwoodman5** [comment]
@holdenhinkle Yes, if we can fix for screen-readers without changing the placement/design right now, that would be ideal. 

Moving the button to the bottom of main has not been part of the conversation until now. We have been considering adding a link within the footer for the Feedback survey (and removing the button from where it is today). However, since this new idea is on the table, I'd like to be able to clearly explain it to our VA Digital Experience lead to decide. If place at the bottom of `main`, would the Feedback button sit left of the `Back to top` button? Thanks.

**Comment by joshkimux** [comment]
Couple of things to consider:
- This fix will only apply to desktop screen readers, not mobile screen readers; most pages may be inaccessible to mobile screen readers until it is no longer sticky on the page.
- Placing the feedback survey in the footer removes it from the context of the content of the page itself.
- It would be more discoverable and accessible to screen readers if it was not designed as an independent button. I'd recommend exploring alternatives with a designer and an accessibility specialist.

**Comment by jwoodman5** [comment]
@joshkimux Re: bullet #3 - precisely why we aren't interested making a quick decision. We have had discussions with designers and accessibility specialists (and others) on our program and will continue. There are plenty of considerations beyond what is listed. If we need assistance, we will reach out. Thanks.

**Comment by holdenhinkle** [comment]
> would the Feedback button sit left of the Back to top button? Thanks.
If we only move the button to `main` and don't change any of the `CSS` (styling), then the button will remain floating on the right side of the screen where it is now. Placing it in `main` will only make it easier for screen readers to tab to the button.

**Comment by holdenhinkle** [comment]
Modified the script so it only moves the Feedback button to the end of `<main>` and does not un-float it or change its appearance in any way.

Moved the script to `src/platform/site-wide/medallia-feedback-button.js`.

**Comment by holdenhinkle** [comment]
I just submitted the PR.

**Comment by jwoodman5** [comment]
> > would the Feedback button sit left of the Back to top button? Thanks.
> 
> If we only move the button to `main` and don't change any of the `CSS` (styling), then the button will remain floating on the right side of the screen where it is now. Placing it in `main` will only make it easier for screen readers to tab to the button.

Awesome!
